File: internal/httpx/reply_feed_context_test.go
package httpx

import (
	"strings"
	"testing"

	"ptxt-nstr/internal/nostrx"
)

func hex64(repeated2 string) string {
	return strings.Repeat(repeated2, 32)
}

func TestIsFeedThreadReply(t *testing.T) {
	root := hex64("aa")
	parent := hex64("bb")
	replyID := hex64("cc")
	author := hex64("01")

	replyTags := [][]string{
		{"e", root, "", "root"},
		{"e", parent, "", "reply"},
		{"p", hex64("02"), ""},
	}
	reply := nostrx.Event{
		ID:        replyID,
		PubKey:    author,
		Kind:      nostrx.KindTextNote,
		Tags:      replyTags,
		Content:   "hi",
	}

	if !isFeedThreadReply(reply) {
		t.Fatal("expected reply with root+reply e-tags to be a thread reply")
	}

	rootNote := nostrx.Event{
		ID:        root,
		PubKey:    author,
		Kind:      nostrx.KindTextNote,
		Tags:      nil,
		Content:   "root",
	}
	if isFeedThreadReply(rootNote) {
		t.Fatal("root note without e-tags should not be a thread reply")
	}

	quote := reply
	quote.Tags = append(append([][]string(nil), replyTags...), []string{"q", parent})
	if !isFeedThreadReply(quote) {
		t.Fatal("quote with e-tags is still a thread reply structurally")
	}
	if replyContextVisible(quote) {
		t.Fatal("quote post should not show reply feed context row")
	}
}

func TestReplyContextVisible(t *testing.T) {
	root := hex64("aa")
	parent := hex64("bb")
	replyID := hex64("cc")
	ev := nostrx.Event{
		ID:     replyID,
		PubKey: hex64("01"),
		Kind:   nostrx.KindTextNote,
		Tags: [][]string{
			{"e", root, "", "root"},
			{"e", parent, "", "reply"},
		},
	}
	if !replyContextVisible(ev) {
		t.Fatal("expected plain reply to show context")
	}
}

func TestReplyContextTargetsDedupesAndSkipsAuthor(t *testing.T) {
	author := hex64("01")
	other := hex64("02")
	ev := nostrx.Event{
		PubKey: author,
		Tags: [][]string{
			{"p", other, ""},
			{"p", author, ""},
			{"p", other, ""},
		},
	}
	got := replyContextTargets(ev)
	if len(got) != 1 || got[0] != other {
		t.Fatalf("got %#v, want single pubkey %q", got, other)
	}
}

func TestReplyContextHTML(t *testing.T) {
	root := hex64("aa")
	parent := hex64("bb")
	replyID := hex64("cc")
	author := hex64("01")
	bob := hex64("02")
	carol := hex64("03")
	dave := hex64("04")
	eve := hex64("05")

	profiles := map[string]nostrx.Profile{
		bob:   {PubKey: bob, Display: "Bob"},
		carol: {PubKey: carol, Display: "Carol"},
		dave:  {PubKey: dave, Display: "Dave"},
		eve:   {PubKey: eve, Display: "Eve"},
	}

	t.Run("two mentions no tail", func(t *testing.T) {
		ev := nostrx.Event{
			ID:     replyID,
			PubKey: author,
			Kind:   nostrx.KindTextNote,
			Tags: [][]string{
				{"e", root, "", "root"},
				{"e", parent, "", "reply"},
				{"p", bob, ""},
				{"p", carol, ""},
			},
		}
		html := string(replyContextHTML(profiles, ev))
		if !strings.Contains(html, "Replying to ") {
			t.Fatalf("missing lead: %q", html)
		}
		if !strings.Contains(html, `href="/u/`+bob+`"`) || !strings.Contains(html, `@Bob`) {
			t.Fatalf("want bob link: %q", html)
		}
		if !strings.Contains(html, `href="/u/`+carol+`"`) || !strings.Contains(html, `@Carol`) {
			t.Fatalf("want carol link: %q", html)
		}
		if strings.Contains(html, "other") {
			t.Fatalf("unexpected others tail: %q", html)
		}
	})

	t.Run("and N others", func(t *testing.T) {
		ev := nostrx.Event{
			ID:     replyID,
			PubKey: author,
			Kind:   nostrx.KindTextNote,
			Tags: [][]string{
				{"e", root, "", "root"},
				{"e", parent, "", "reply"},
				{"p", bob, ""},
				{"p", carol, ""},
				{"p", dave, ""},
				{"p", eve, ""},
			},
		}
		html := string(replyContextHTML(profiles, ev))
		if !strings.Contains(html, "and 2 others") {
			t.Fatalf("want and 2 others: %q", html)
		}
	})

	t.Run("fallback thread link when no p tags", func(t *testing.T) {
		ev := nostrx.Event{
			ID:     replyID,
			PubKey: author,
			Kind:   nostrx.KindTextNote,
			Tags: [][]string{
				{"e", root, "", "root"},
				{"e", parent, "", "reply"},
			},
		}
		html := string(replyContextHTML(nil, ev))
		if !strings.Contains(html, `href="/thread/`+parent+`"`) || !strings.Contains(html, ">thread</a>") {
			t.Fatalf("want parent thread link: %q", html)
		}
	})

	t.Run("quote post empty", func(t *testing.T) {
		ev := nostrx.Event{
			ID:     replyID,
			PubKey: author,
			Kind:   nostrx.KindTextNote,
			Tags: [][]string{
				{"e", root, "", "root"},
				{"e", parent, "", "reply"},
				{"q", parent},
				{"p", bob, ""},
			},
		}
		if got := string(replyContextHTML(profiles, ev)); got != "" {
			t.Fatalf("want empty for quote: %q", got)
		}
	})
}

func TestRepostContextHTML(t *testing.T) {
	alice := hex64("0a")
	ev := nostrx.Event{
		ID:        hex64("f0"),
		PubKey:    alice,
		Kind:      nostrx.KindRepost,
		Tags:      [][]string{{"e", hex64("aa"), ""}},
		Content:   "",
	}
	profiles := map[string]nostrx.Profile{
		alice: {PubKey: alice, Display: "Alice"},
	}
	html := string(repostContextHTML(profiles, ev))
	if !strings.Contains(html, "reposted") {
		t.Fatalf("missing reposted: %q", html)
	}
	if !strings.Contains(html, "Alice") {
		t.Fatalf("missing display name: %q", html)
	}
	if !strings.Contains(html, `note-feed-context-repost-inner`) {
		t.Fatalf("missing inner span: %q", html)
	}

	non := nostrx.Event{Kind: nostrx.KindTextNote, PubKey: alice}
	if got := string(repostContextHTML(profiles, non)); got != "" {
		t.Fatalf("kind 1 should not emit repost banner: %q", got)
	}
}

File: web/static/js/reply-feed-context.test.js
import assert from "node:assert/strict";
import { describe, it } from "node:test";
import { profilePath } from "./relay-utils.js";

import {
  isFeedThreadReply,
  replyContextHTML,
  replyContextTargets,
  replyContextVisible,
  repostContextHTML,
} from "./reply-feed-context.js";

function hex64(repeated2) {
  return repeated2.repeat(32);
}

describe("reply-feed-context", () => {
  it("detects thread replies and hides quote posts", () => {
    const root = hex64("aa");
    const parent = hex64("bb");
    const replyID = hex64("cc");
    const author = hex64("01");

    const reply = {
      id: replyID,
      pubkey: author,
      kind: 1,
      tags: [
        ["e", root, "", "root"],
        ["e", parent, "", "reply"],
        ["p", hex64("02"), ""],
      ],
      content: "hi",
    };
    assert.equal(isFeedThreadReply(reply), true);
    assert.equal(replyContextVisible(reply), true);

    const rootNote = {
      id: root,
      pubkey: author,
      kind: 1,
      tags: [],
      content: "root",
    };
    assert.equal(isFeedThreadReply(rootNote), false);

    const quote = {
      ...reply,
      tags: [...reply.tags, ["q", parent]],
    };
    assert.equal(isFeedThreadReply(quote), true);
    assert.equal(replyContextVisible(quote), false);
  });

  it("dedupes reply targets and skips author", () => {
    const author = hex64("01");
    const other = hex64("02");
    const targets = replyContextTargets({
      pubkey: author,
      tags: [
        ["p", other, ""],
        ["p", author, ""],
        ["p", other, ""],
      ],
    });
    assert.deepEqual(targets, [other]);
  });

  it("builds reply context HTML with mentions", () => {
    const root = hex64("aa");
    const parent = hex64("bb");
    const bob = hex64("02");
    const carol = hex64("03");
    const profiles = {
      [bob]: { pubkey: bob, display_name: "Bob" },
      [carol]: { pubkey: carol, display_name: "Carol" },
    };
    const html = replyContextHTML(
      {
        id: hex64("cc"),
        pubkey: hex64("01"),
        kind: 1,
        tags: [
          ["e", root, "", "root"],
          ["e", parent, "", "reply"],
          ["p", bob, ""],
          ["p", carol, ""],
        ],
      },
      profiles,
    );
    assert.match(html, /Replying to/);
    assert.match(html, /@Bob/);
    assert.match(html, /@Carol/);
    assert.match(html, new RegExp(`href="${profilePath(bob).replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}"`));
  });

  it("falls back to thread link without p tags", () => {
    const root = hex64("aa");
    const parent = hex64("bb");
    const html = replyContextHTML({
      id: hex64("cc"),
      pubkey: hex64("01"),
      kind: 1,
      tags: [
        ["e", root, "", "root"],
        ["e", parent, "", "reply"],
      ],
    });
    assert.match(html, new RegExp(`href="/thread/${parent}"`));
    assert.match(html, />thread</);
  });

  it("builds repost banner HTML", () => {
    const alice = hex64("0a");
    const html = repostContextHTML(
      {
        id: hex64("f0"),
        pubkey: alice,
        kind: 6,
        tags: [["e", hex64("aa"), ""]],
        content: "",
      },
      { [alice]: { pubkey: alice, display_name: "Alice" } },
    );
    assert.match(html, /Alice reposted/);
    assert.match(html, /note-feed-context-repost-inner/);
  });
});
